File: apps/ios/src/services/signaling/signalingClient.ts
import { backendEnvironment } from '../../config/backendEnvironment';
import {
  parseIncomingMessage,
  serializeOutgoingMessage,
  type IncomingMessage,
  type OutgoingMessage,
} from '../../types/protocol';

export type ConnectionStatus = 'disconnected' | 'connecting' | 'connected';

/** How often a heartbeat `session.ping` goes out while connected. */
const HEARTBEAT_INTERVAL_MS = 20_000;
/** Silence longer than this means the socket is dead even if the OS hasn't noticed yet —
 * common right after a Wi-Fi <-> cellular switch, where a TCP connection can hang for minutes. */
const STALE_AFTER_MS = 45_000;
/** How long `probe()` waits for any reply before declaring the socket dead. */
const PROBE_TIMEOUT_MS = 5_000;

/** Pulls the fields that actually matter off a WS close/error event without risking a
 * circular-reference crash from JSON.stringify on the event's `.target` back-reference. */
function safeStringify(event: unknown): string {
  const e = event as { code?: unknown; reason?: unknown; message?: unknown; type?: unknown };
  return JSON.stringify({ code: e?.code, reason: e?.reason, message: e?.message, type: e?.type });
}

/**
 * Authenticated WebSocket connection to the backend's `/ws` signaling
 * endpoint (see docs/protocol.md). Carries pairing and presence messages
 * only — terminal/screen/input data goes over WebRTC once Phase 4 exists,
 * never through here. Mirrors apps/mac/RemoteMac's SignalingClient.swift.
 *
 * React Native's WebSocket accepts a non-standard third `options.headers`
 * argument, which is how the Authorization bearer token reaches the
 * upgrade request — the same mechanism the backend expects from the Mac
 * client (docs/protocol.md's "Connecting to /ws").
 */
export class SignalingClient {
  private socket: WebSocket | null = null;
  private pendingPing: { resolve: (receivedAt: number) => void; reject: (error: Error) => void } | null = null;
  private pingTimeout: ReturnType<typeof setTimeout> | null = null;
  private heartbeat: ReturnType<typeof setInterval> | null = null;
  private probeTimeout: ReturnType<typeof setTimeout> | null = null;
  private lastReceivedAt = 0;
  /** Set once this socket has reported 'disconnected', so close + error (which RN fires
   * back to back) and a heartbeat-detected death only ever report it once. */
  private closed = false;

  onMessage: ((message: IncomingMessage) => void) | null = null;
  onStatusChange: ((status: ConnectionStatus) => void) | null = null;
  /** Diagnostic-only: raw open/close/error event details, since Release-build console.log
   * isn't reachable via CLI from a physical device. Remove once WS connectivity is fixed. */
  onDebug: ((detail: string) => void) | null = null;

  constructor(
    private readonly deviceId: string,
    private readonly accessToken: string,
    private readonly baseURL: string = backendEnvironment.webSocketBaseURL,
  ) {}

  connect(): void {
    this.onStatusChange?.('connecting');

    const url = `${this.baseURL}?deviceId=${encodeURIComponent(this.deviceId)}`;
    this.onDebug?.(`connect() url=${url} tokenPrefix=${this.accessToken.slice(0, 12)}`);
    const socket = new WebSocket(url, undefined, {
      headers: { Authorization: `Bearer ${this.accessToken}` },
    });

    socket.onopen = () => {
      this.onDebug?.('onopen');
      this.lastReceivedAt = Date.now();
      this.startHeartbeat();
      this.onStatusChange?.('connected');
    };
    socket.onclose = (event: unknown) => {
      this.onDebug?.(`onclose ${safeStringify(event)}`);
      this.markClosed(new Error('Connection closed'));
    };
    socket.onerror = (event: unknown) => {
      this.onDebug?.(`onerror ${safeStringify(event)}`);
      this.markClosed(new Error('Connection error'));
    };
    socket.onmessage = (event) => {
      this.lastReceivedAt = Date.now();
      this.clearProbe();
      this.handleMessage(String(event.data));
    };

    this.socket = socket;
  }

  /** Intentional close: reports nothing, so the owner's reconnect logic doesn't fire. */
  disconnect(): void {
    this.closed = true;
    this.detachAndClose();
    this.failPendingPing(new Error('Disconnected'));
  }

  /**
   * Checks the socket is actually alive rather than trusting `readyState` — after the app
   * was backgrounded or the network changed, a socket can still claim OPEN while nothing
   * gets through. No reply within a few seconds is treated as a drop.
   */
  probe(): void {
    if (this.closed || !this.socket || this.socket.readyState !== WebSocket.OPEN) return;
    if (this.probeTimeout) return;
    this.send({ type: 'session.ping', payload: { sentAt: Date.now() } });
    this.probeTimeout = setTimeout(() => {
      this.probeTimeout = null;
      this.markClosed(new Error('Connection timed out'));
    }, PROBE_TIMEOUT_MS);
  }

  private startHeartbeat(): void {
    this.stopHeartbeat();
    this.heartbeat = setInterval(() => {
      if (Date.now() - this.lastReceivedAt > STALE_AFTER_MS) {
        this.markClosed(new Error('Connection timed out'));
        return;
      }
      this.send({ type: 'session.ping', payload: { sentAt: Date.now() } });
    }, HEARTBEAT_INTERVAL_MS);
  }

  private stopHeartbeat(): void {
    if (this.heartbeat) {
      clearInterval(this.heartbeat);
      this.heartbeat = null;
    }
  }

  private clearProbe(): void {
    if (this.probeTimeout) {
      clearTimeout(this.probeTimeout);
      this.probeTimeout = null;
    }
  }

  /** Unintentional loss of the connection — reported exactly once as 'disconnected'. */
  private markClosed(error: Error): void {
    if (this.closed) return;
    this.closed = true;
    this.detachAndClose();
    this.failPendingPing(error);
    this.onStatusChange?.('disconnected');
  }

  private detachAndClose(): void {
    this.stopHeartbeat();
    this.clearProbe();
    const socket = this.socket;
    this.socket = null;
    if (!socket) return;
    socket.onopen = null;
    socket.onclose = null;
    socket.onerror = null;
    socket.onmessage = null;
    socket.close();
  }

  send(message: OutgoingMessage): void {
    if (!this.socket || this.socket.readyState !== WebSocket.OPEN) {
      this.onDebug?.(`send() DROPPED type=${message.type} readyState=${this.socket?.readyState}`);
      return;
    }
    this.socket.send(serializeOutgoingMessage(message));
  }

  /**
   * Sends `session.ping` and resolves with the receive timestamp once the
   * matching `session.pong` arrives (or rejects after a 5s timeout). Only
   * one ping should be in flight at a time — used for the "test connection"
   * action, a single user-initiated check.
   */
  ping(sentAt: number): Promise<number> {
    this.send({ type: 'session.ping', payload: { sentAt } });
    return new Promise<number>((resolve, reject) => {
      this.pendingPing = { resolve, reject };
      this.pingTimeout = setTimeout(() => {
        this.failPendingPing(new Error('The connection did not respond in time.'));
      }, 5000);
    });
  }

  private failPendingPing(error: Error): void {
    if (this.pingTimeout) {
      clearTimeout(this.pingTimeout);
      this.pingTimeout = null;
    }
    this.pendingPing?.reject(error);
    this.pendingPing = null;
  }

  private handleMessage(raw: string): void {
    const message = parseIncomingMessage(raw);
    if (message.type === 'session.pong') {
      if (this.pingTimeout) {
        clearTimeout(this.pingTimeout);
        this.pingTimeout = null;
      }
      this.pendingPing?.resolve(Date.now());
      this.pendingPing = null;
    }
    this.onMessage?.(message);
  }
}

File: apps/mac/RemoteMac/Networking/Signaling/SignalingClient.swift
import Foundation

/// `print` is a synchronous stdout write that still runs in Release builds — fine for occasional
/// events, but this fires on every signaling message, which only makes sense as a debug aid.
@inline(__always)
private func debugLog(_ message: @autoclosure () -> String) {
    #if DEBUG
    print(message())
    #endif
}

enum ConnectionStatus: Equatable {
    case disconnected
    case connecting
    case connected
}

enum SignalingError: Error, LocalizedError {
    case timeout
    case notConnected

    var errorDescription: String? {
        switch self {
        case .timeout: return "The connection did not respond in time."
        case .notConnected: return "Not connected to the backend."
        }
    }
}

/// Authenticated WebSocket connection to the backend's `/ws` signaling
/// endpoint (see docs/protocol.md). Carries pairing and presence messages
/// only — terminal/screen/input data goes over WebRTC once Phase 4 exists,
/// never through here.
///
/// Not actor-isolated: `URLSession` is configured with `delegateQueue: .main`
/// so every callback below already lands on the main thread, and this class
/// is only ever driven from `AppState` (`@MainActor`). Marked `@unchecked
/// Sendable` to reflect that externally-enforced invariant to the compiler.
final class SignalingClient: NSObject, @unchecked Sendable {
    var onMessage: ((ProtocolMessage) -> Void)?
    var onStatusChange: ((ConnectionStatus) -> Void)?

    private let deviceId: String
    private let accessToken: String
    private let baseURL: URL
    private var session: URLSession!
    private var task: URLSessionWebSocketTask?
    private var pendingPing: CheckedContinuation<Double, Error>?
    private var heartbeatTimer: Timer?
    /// Set once this connection has reported `.disconnected` — URLSession can surface the same
    /// drop through several callbacks (receive failure, didClose, didCompleteWithError), and
    /// the owner's reconnect logic should only hear about it once.
    private var closed = false
    private var awaitingPong = false

    /// How often to send a WebSocket ping while connected, and how long to wait for its pong.
    /// A dead connection (e.g. after the network changed) otherwise isn't noticed for minutes.
    private static let heartbeatInterval: TimeInterval = 20
    private static let heartbeatTimeout: UInt64 = 10_000_000_000

    init(deviceId: String, accessToken: String, baseURL: URL = BackendEnvironment.current.webSocketBaseURL) {
        self.deviceId = deviceId
        self.accessToken = accessToken
        self.baseURL = baseURL
        super.init()
        self.session = URLSession(configuration: .default, delegate: self, delegateQueue: .main)
    }

    func connect() {
        guard var components = URLComponents(url: baseURL, resolvingAgainstBaseURL: false) else { return }
        components.queryItems = [URLQueryItem(name: "deviceId", value: deviceId)]
        guard let url = components.url else { return }

        var request = URLRequest(url: url)
        request.setValue("Bearer \(accessToken)", forHTTPHeaderField: "Authorization")

        debugLog("[RM-WS] connect() url=\(url) deviceId=\(deviceId) tokenPrefix=\(accessToken.prefix(12))")
        onStatusChange?(.connecting)
        let task = session.webSocketTask(with: request)
        self.task = task
        task.resume()
        listen()
    }

    /// Intentional close: reports nothing, so the owner's reconnect logic doesn't fire.
    func disconnect() {
        closed = true
        stopHeartbeat()
        task?.cancel(with: .goingAway, reason: nil)
        task = nil
        session.invalidateAndCancel()
        failPendingPing(with: SignalingError.notConnected)
    }

    /// Checks the connection is actually alive — used after the Mac wakes or the network
    /// changes, when the socket can look open while nothing gets through.
    func probe() {
        guard !closed, let task, !awaitingPong else { return }
        awaitingPong = true
        task.sendPing { [weak self] error in
            DispatchQueue.main.async {
                self?.awaitingPong = false
                if error != nil { self?.markClosed() }
            }
        }
        Task { @MainActor [weak self] in
            try? await Task.sleep(nanoseconds: Self.heartbeatTimeout)
            guard let self, self.awaitingPong else { return }
            self.markClosed()
        }
    }

    private func startHeartbeat() {
        stopHeartbeat()
        heartbeatTimer = Timer.scheduledTimer(withTimeInterval: Self.heartbeatInterval, repeats: true) { [weak self] _ in
            self?.probe()
        }
    }

    private func stopHeartbeat() {
        heartbeatTimer?.invalidate()
        heartbeatTimer = nil
    }

    /// Unintentional loss of the connection — reported exactly once as `.disconnected`.
    private func markClosed() {
        guard !closed else { return }
        closed = true
        stopHeartbeat()
        task?.cancel(with: .goingAway, reason: nil)
        task = nil
        session.invalidateAndCancel()
        failPendingPing(with: SignalingError.notConnected)
        onStatusChange?(.disconnected)
    }

    func send(_ message: OutgoingProtocolMessage) {
        guard let task else { return }
        guard let data = try? message.encoded() else {
            // Encoding a fixed, well-typed payload should never fail; if it does there is
            // nothing a caller could usefully retry.
            assertionFailure("Failed to encode a well-typed outgoing signaling message")
            return
        }
        task.send(.data(data)) { _ in }
    }

    /// Sends `session.ping` and resolves once the matching `session.pong` arrives (or after a
    /// 5s timeout). Only one ping should be in flight at a time — used for the onboarding
    /// "test connection" step, which is a single user-initiated action.
    func ping(sentAt: Double) async throws -> Double {
        send(.sessionPing(SessionPingPayload(sentAt: sentAt)))
        return try await withCheckedThrowingContinuation { continuation in
            self.pendingPing = continuation
            Task { @MainActor [weak self] in
                try? await Task.sleep(nanoseconds: 5_000_000_000)
                self?.failPendingPing(with: SignalingError.timeout)
            }
        }
    }

    private func failPendingPing(with error: Error) {
        pendingPing?.resume(throwing: error)
        pendingPing = nil
    }

    private func listen() {
        task?.receive { [weak self] result in
            guard let self else { return }
            switch result {
            case .failure(let error):
                debugLog("[RM-WS] receive failure: \(error)")
                self.markClosed()
            case .success(let message):
                switch message {
                case .data(let data):
                    self.handle(data)
                case .string(let text):
                    self.handle(Data(text.utf8))
                @unknown default:
                    break
                }
                self.listen()
            }
        }
    }

    private func handle(_ data: Data) {
        guard let message = try? ProtocolMessage.decode(from: data) else {
            debugLog("[RM-WS] recv UNDECODABLE: \(String(data: data, encoding: .utf8) ?? "<binary>")")
            return
        }
        debugLog("[RM-WS] recv \(message)")
        if case .sessionPong = message {
            pendingPing?.resume(returning: Date().timeIntervalSince1970)
            pendingPing = nil
        }
        onMessage?(message)
    }
}

extension SignalingClient: URLSessionWebSocketDelegate {
    func urlSession(
        _ session: URLSession,
        webSocketTask: URLSessionWebSocketTask,
        didOpenWithProtocol protocol: String?
    ) {
        debugLog("[RM-WS] didOpenWithProtocol \(`protocol` ?? "nil")")
        guard !closed else { return }
        startHeartbeat()
        onStatusChange?(.connected)
    }

    func urlSession(
        _ session: URLSession,
        webSocketTask: URLSessionWebSocketTask,
        didCloseWith closeCode: URLSessionWebSocketTask.CloseCode,
        reason: Data?
    ) {
        let reasonText = reason.flatMap { String(data: $0, encoding: .utf8) } ?? "nil"
        debugLog("[RM-WS] didCloseWith code=\(closeCode.rawValue) reason=\(reasonText)")
        markClosed()
    }
}

extension SignalingClient: URLSessionTaskDelegate {
    func urlSession(_ session: URLSession, task: URLSessionTask, didCompleteWithError error: Error?) {
        guard let error else { return }
        debugLog("[RM-WS] didCompleteWithError \(error)")
        if let response = task.response as? HTTPURLResponse {
            debugLog("[RM-WS] didCompleteWithError httpStatus=\(response.statusCode) headers=\(response.allHeaderFields)")
        }
        markClosed()
    }
}
